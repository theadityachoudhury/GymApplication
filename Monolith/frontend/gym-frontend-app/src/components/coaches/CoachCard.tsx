import { coachesType } from '@/types/coaches/coaches.type';
import Button from '../common/ui/CustomButton';
import { useNavigate } from 'react-router';

function CoachCard({ coach }: { coach: coachesType }) {
	const {
		name,
		summary: title,
		motivationPitch: description,
		rating,
		imageUrl: image,
	} = coach;
	const navigator = useNavigate();
	return (
		<div className="flex flex-col rounded-2xl h-full bg-white shadow-lg overflow-hidden flex-1 min-w-[250px] max-w-[300px]">
			{/* Image */}
			<div className="w-full aspect-[4/3]">
				<img
					src={image || '/placeholder.svg'}
					alt={`${name}`}
					className="w-full h-full object-cover object-center"
				/>
			</div>

			{/* Content */}
			<div className="p-5 flex flex-col flex-1 justify-between">
				<div className="flex justify-between items-start mb-1">
					<h2 className="text-lg font-semibold text-gray-800">
						{name}
					</h2>
					<div className="flex items-center gap-1">
						<span className="text-sm font-semibold text-gray-700">
							{rating.toFixed(2)}
						</span>
						<svg
							width="16"
							height="16"
							viewBox="0 0 24 24"
							fill="currentColor"
							xmlns="http://www.w3.org/2000/svg"
							className="text-yellow-400"
						>
							<path d="M12 2L15.09 8.26L22 9.27L17 14.14L18.18 21.02L12 17.77L5.82 21.02L7 14.14L2 9.27L8.91 8.26L12 2Z" />
						</svg>
					</div>
				</div>

				<h3 className="text-sm text-gray-500 mb-3">{title}</h3>

				<p className="text-sm text-gray-600 mb-5 leading-relaxed line-clamp-3">
					{description}
				</p>

				<Button onClick={() => navigator(`/coaches/${coach.id}`)}>
					Book Workout
				</Button>
			</div>
		</div>
	);
}

export default CoachCard;
