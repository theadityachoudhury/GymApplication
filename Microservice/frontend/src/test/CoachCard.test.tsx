import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom'; // Import from react-router-dom, not react-router
import CoachCard from '../components/coaches/CoachCard';

// Mock react-router's useNavigate
vi.mock('react-router', () => ({
	useNavigate: () => mockNavigate,
}));

// Create the mock navigate function
const mockNavigate = vi.fn();

// Mock coach data
const mockCoach = {
	id: '123',
	name: 'John Doe',
	summary: 'Fitness Expert',
	motivationPitch:
		'I will help you achieve your fitness goals with personalized training programs.',
	rating: 4.85,
	imageUrl: 'https://example.com/coach-image.jpg',
	specializations: ['Yoga', 'Strength Training'],
};

// Mock coach without image
const mockCoachNoImage = {
	...mockCoach,
	imageUrl: '',
};

describe('CoachCard Component', () => {
	// Reset mocks before each test
	beforeEach(() => {
		mockNavigate.mockReset();
	});

	// Helper function to render component within router context
	const renderCoachCard = (coach = mockCoach) => {
		return render(
			<MemoryRouter>
				<CoachCard coach={coach} />
			</MemoryRouter>
		);
	};

	it('renders coach information correctly', () => {
		renderCoachCard();

		// Check if name is displayed
		expect(screen.getByText('John Doe')).toBeInTheDocument();

		// Check if title (summary) is displayed
		expect(screen.getByText('Fitness Expert')).toBeInTheDocument();

		// Check if description (motivationPitch) is displayed
		expect(
			screen.getByText(
				'I will help you achieve your fitness goals with personalized training programs.'
			)
		).toBeInTheDocument();

		// Check if rating is displayed and formatted correctly
		expect(screen.getByText('4.85')).toBeInTheDocument();
	});

	it('displays the coach image when provided', () => {
		renderCoachCard();

		const image = screen.getByAltText('John Doe');
		expect(image).toBeInTheDocument();
		expect(image.getAttribute('src')).toBe(
			'https://example.com/coach-image.jpg'
		);
	});

	it('displays a placeholder image when no image is provided', () => {
		renderCoachCard(mockCoachNoImage);

		const image = screen.getByAltText('John Doe');
		expect(image).toBeInTheDocument();
		expect(image.getAttribute('src')).toBe('/placeholder.svg');
	});

	it('navigates to coach detail page when "Book Workout" button is clicked', async () => {
		renderCoachCard();

		const bookButton = screen.getByText('Book Workout');
		expect(bookButton).toBeInTheDocument();

		// Click the button
		await userEvent.click(bookButton);

		// Check if navigation was called with the correct path
		expect(mockNavigate).toHaveBeenCalledTimes(1);
		expect(mockNavigate).toHaveBeenCalledWith('/coaches/123');
	});

	it('renders the star icon for rating', () => {
		renderCoachCard();

		// Check if the SVG star icon is rendered
		const starIcon = document.querySelector('svg');
		expect(starIcon).toBeInTheDocument();
		expect(starIcon?.classList.contains('text-yellow-400')).toBe(true);
	});

	it('applies correct styling to the card', () => {
		const { container } = renderCoachCard();

		// Check main card container
		const cardContainer = container.firstChild;
		expect(cardContainer).toHaveClass('flex');
		expect(cardContainer).toHaveClass('flex-col');
		expect(cardContainer).toHaveClass('rounded-2xl');
		expect(cardContainer).toHaveClass('bg-white');
		expect(cardContainer).toHaveClass('shadow-lg');

		// Check image container
		const imageContainer = container.querySelector('.aspect-\\[4\\/3\\]');
		expect(imageContainer).toBeInTheDocument();

		// Check content container
		const contentContainer = container.querySelector('.p-5');
		expect(contentContainer).toBeInTheDocument();
	});

	it('truncates long description text with line-clamp-3', () => {
		const longDescriptionCoach = {
			...mockCoach,
			motivationPitch:
				'This is a very long description that should be truncated. '.repeat(
					10
				),
		};

		renderCoachCard(longDescriptionCoach);

		const description = screen.getByText(/This is a very long description/);
		expect(description).toHaveClass('line-clamp-3');
	});

	it('formats rating to 2 decimal places', () => {
		const coachWithUglyRating = {
			...mockCoach,
			rating: 4.8333333,
		};

		renderCoachCard(coachWithUglyRating);

		// Should display as 4.83, not 4.8333333
		expect(screen.getByText('4.83')).toBeInTheDocument();
	});

	it('renders a Button component for booking', () => {
		renderCoachCard();

		const button = screen.getByRole('button', { name: 'Book Workout' });
		expect(button).toBeInTheDocument();
	});

	// Snapshot test
	it('matches snapshot', () => {
		const { container } = renderCoachCard();
		expect(container).toMatchSnapshot();
	});
});
