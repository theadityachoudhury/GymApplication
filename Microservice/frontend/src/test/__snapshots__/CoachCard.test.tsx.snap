// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoachCard Component > matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col rounded-2xl h-full bg-white shadow-lg overflow-hidden flex-1 min-w-[250px] max-w-[300px]"
  >
    <div
      class="w-full aspect-[4/3]"
    >
      <img
        alt="John Doe"
        class="w-full h-full object-cover object-center"
        src="https://example.com/coach-image.jpg"
      />
    </div>
    <div
      class="p-5 flex flex-col flex-1 justify-between"
    >
      <div
        class="flex justify-between items-start mb-1"
      >
        <h2
          class="text-lg font-semibold text-gray-800"
        >
          John Doe
        </h2>
        <div
          class="flex items-center gap-1"
        >
          <span
            class="text-sm font-semibold text-gray-700"
          >
            4.85
          </span>
          <svg
            class="text-yellow-400"
            fill="currentColor"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2L15.09 8.26L22 9.27L17 14.14L18.18 21.02L12 17.77L5.82 21.02L7 14.14L2 9.27L8.91 8.26L12 2Z"
            />
          </svg>
        </div>
      </div>
      <h3
        class="text-sm text-gray-500 mb-3"
      >
        Fitness Expert
      </h3>
      <p
        class="text-sm text-gray-600 mb-5 leading-relaxed line-clamp-3"
      >
        I will help you achieve your fitness goals with personalized training programs.
      </p>
      <button
        class="
        rounded-md font-medium transition-all flex items-center justify-center disabled:cursor-not-allowed
        bg-primary-green text-black hover:bg-primary-white border border-primary-green hover:border-black cursor-pointer
        text-base py-3 px-4
        
        
        
      "
        type="button"
      >
        Book Workout
      </button>
    </div>
  </div>
</div>
`;
